libraries/AP_HAL_Linux/PWM_Sysfs.cpp

```diff
@@ -0,0 +1,218 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+ * Copyright (C) 2015  Intel Corporation. All rights reserved.
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+
+#include "PWM_Sysfs.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <AP_Math/AP_Math.h>
+
+static const AP_HAL::HAL &hal = AP_HAL::get_HAL();
+
+namespace Linux {
+
+/* Trick to use minimum stack space for each of the params */
+union pwm_params {
+    char export_gpio[sizeof("XXX/export")];
+    char duty_cycle[sizeof("XXX/pwmXXX/duty_cycle")];
+    char enable[sizeof("XXX/pwmXXX/enable")];
+    char period[sizeof("XXX/pwmXXX/period")];
+    char polarity[sizeof("XXX/pwmXXX/polarity")];
+};
+#define PWM_BASE_PATH "/sys/class/pwm/pwmchip"
+#define PWM_PATH_MAX (sizeof(PWM_BASE_PATH) + sizeof(pwm_params) - 1)
+
+PWM_Sysfs::PWM_Sysfs(uint8_t chip, uint8_t channel)
+    : _chip(chip)
+    , _channel(channel)
+    , _duty_cycle_fd(-1)
+    , _nsec_duty_cycle_value(0)
+{
+    char path[PWM_PATH_MAX];
+    int r;
+
+    r = snprintf(path, sizeof(path), PWM_BASE_PATH "%u/export", _chip);
+    if (r < 0) {
```

Comment:
I added the check because it doesn't hurt, however notice we always have enough space since we pre calculated it.

Nonetheless I'll add it here, too.